File: azure-functions-ts/src/shared/geminiService.ts
/**
 * Gemini 2.5 Flash AI Service — with Redis caching.
 * Endpoint: ai.sumopod.com
 * Strict output: no markdown symbols, numbered bullet points only.
 */
import { getSecret } from "./keyVault";
import { getCachedResponse, setCachedResponse } from "./redisCache";

const SYSTEM_PROMPT = [
  "You are GeoSupplyGuard AI, a supply chain intelligence analyst for warehouse and inventory management.",
  "",
  "OUTPUT RULES:",
  "1. Always respond in English.",
  "2. Use numbered points (1. 2. 3.) for lists.",
  "3. NEVER use markdown symbols: no asterisks (*), no hyphens (-), no underscores (_), no hash (#), no backticks.",
  "4. Use plain numbered lists only. No bullet characters of any kind.",
  "5. Keep responses thorough but concise (300-500 words max).",
  "6. Focus on actionable recommendations with specific numbers.",
  "7. When given CSV data context, reference specific products, stock levels, and sales figures.",
  "8. For section headers use UPPERCASE TEXT followed by a colon, e.g. RISK ASSESSMENT:",
  "9. Always complete your analysis. Never stop mid-sentence.",
].join("\n");

interface GeminiResponse {
  choices?: Array<{
    message?: {
      content?: string;
    };
  }>;
}

// Bump when prompt/format/token settings change to avoid stale cached responses.
const CACHE_VERSION = "v2_2048_tokens";
const REQUEST_TIMEOUT_MS = 15000;
const MAX_RETRIES = 2;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function shouldRetryStatus(status: number): boolean {
  return status === 408 || status === 429 || status >= 500;
}

async function postWithRetry(
  url: string,
  payload: unknown,
  apiKey: string
): Promise<Response> {
  let lastError: Error | null = null;

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);

    try {
      const res = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Authorization": `Bearer ${apiKey}`,
        },
        body: JSON.stringify(payload),
        signal: controller.signal,
      });

      if (res.ok) return res;

      if (!shouldRetryStatus(res.status) || attempt === MAX_RETRIES) {
        return res;
      }

      await sleep(300 * Math.pow(2, attempt));
      continue;
    } catch (err: any) {
      lastError = err;
      if (attempt === MAX_RETRIES) {
        break;
      }
      await sleep(300 * Math.pow(2, attempt));
    } finally {
      clearTimeout(timeout);
    }
  }

  throw new Error(`Gemini API request failed after retries: ${lastError?.message ?? "unknown error"}`);
}

/**
 * Call Gemini 2.5 Flash via Sumopod OpenAI-compatible API, with Redis caching.
 * If an identical query was answered within TTL (4h), return cached result.
 */
export async function callGemini(
  userPrompt: string,
  context: string = ""
): Promise<{ text: string; cached: boolean }> {
  // 1. Check Redis cache
  const cacheKey = `${CACHE_VERSION}::${userPrompt}::${context}`.trim();
  const cached = await getCachedResponse(cacheKey);
  if (cached) {
    return { text: cached, cached: true };
  }

  // 2. Build request (OpenAI-compatible format for Sumopod)
  const apiKey = await getSecret("SUMOPOD-API-KEY", "SUMOPOD_API_KEY");
  let baseUrl = await getSecret("SUMOPOD-BASE-URL", "SUMOPOD_BASE_URL");
  if (!baseUrl.startsWith("http")) baseUrl = `https://${baseUrl}`;

  const payload = {
    model: "gemini/gemini-2.5-flash",
    messages: [
      { role: "system", content: SYSTEM_PROMPT },
      { role: "user", content: context ? `CONTEXT DATA:\n${context}\n\nQUESTION:\n${userPrompt}` : userPrompt },
    ],
    temperature: 0.3,
    max_tokens: 2048,
    top_p: 0.8,
  };

  const url = `${baseUrl}/v1/chat/completions`;

  // 3. Call API with timeout + retries for transient failures
  const res = await postWithRetry(url, payload, apiKey);

  if (!res.ok) {
    const errBody = await res.text();
    throw new Error(`Gemini API error ${res.status}: ${errBody}`);
  }

  const data = (await res.json()) as GeminiResponse;
  const text =
    data.choices?.[0]?.message?.content ??
    "Unable to generate analysis. Please try again.";

  // 4. Cache the response (4 hours TTL)
  await setCachedResponse(cacheKey, text, 4 * 60 * 60);

  return { text, cached: false };
}

File: azure-functions/shared/gemini_service.py
"""
Shared module: Gemini 2.5 Flash AI service.
Calls the custom endpoint at ai.sumopod.com with strict short/bullet-point formatting.
"""

import logging
import httpx
from shared.keyvault import get_secret

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = (
    "You are GeoSupplyGuard AI, a concise geopolitical supply chain risk analyst.\n"
    "Rules:\n"
    "1. Always respond in English.\n"
    "2. Use short bullet points. Each point max 15 words.\n"
    "3. Never use markdown symbols such as asterisks, hyphens, or underscores.\n"
    "4. Use numbered lists (1. 2. 3.) instead of bullet characters.\n"
    "5. Keep total response under 200 words.\n"
    "6. Focus on actionable risk mitigation steps.\n"
    "7. Cite geopolitical events only when directly relevant (e.g., sanctions, conflicts, trade bans).\n"
)


async def call_gemini(user_prompt: str, context: str = "") -> str:
    """
    Send a prompt to Gemini 2.5 Flash via the custom endpoint.
    Returns the plain-text response body.
    """
    api_key = get_secret("SUMOPOD-API-KEY", env_fallback="SUMOPOD_API_KEY")
    base_url = get_secret("SUMOPOD-BASE-URL", env_fallback="SUMOPOD_BASE_URL")
    if not base_url.startswith("http"):
        base_url = f"https://{base_url}"

    full_prompt = f"{SYSTEM_PROMPT}\n\nCONTEXT DATA:\n{context}\n\nUSER QUESTION:\n{user_prompt}"

    payload = {
        "contents": [
            {
                "parts": [
                    {"text": full_prompt}
                ]
            }
        ],
        "generationConfig": {
            "temperature": 0.3,
            "maxOutputTokens": 512,
            "topP": 0.8,
        },
    }

    url = (
        f"{base_url}/v1beta/models/gemini-2.5-flash:generateContent"
        f"?key={api_key}"
    )

    async with httpx.AsyncClient(timeout=30.0) as client:
        resp = await client.post(url, json=payload)
        resp.raise_for_status()
        data = resp.json()

    try:
        text = data["candidates"][0]["content"]["parts"][0]["text"]
    except (KeyError, IndexError):
        logger.error("Unexpected Gemini response structure: %s", data)
        text = "Unable to generate analysis. Please try again."

    return text
